### PR TITLE
Guard on inner tensor names for traceable wrapper subclasses

### LINKED ISSUE #129601: dynamo should recompile when a tensor subclass's inner tensor changes
The problem is that if a tensor subclass wants to be able to hold an "optional" inner tensor, then:

(1) we will specialize our graph on whether the inner tensor is set or not
(2) we therefore must guard on whether or not it is set.

We are not guarding, however, so we end up incorrectly re-using an existing graph when we should be recompiling.

The easiest fix is probably that when dynamo sees a GraphArg that is a tensor subclass, it should guard on the exact attribute names that are returned by `__tensor_flatten__`. We expect that as a subclass author, you should write your __tensor_flatten__ to return a list of attributes corresponding only to the current valid inner tensors (and if your inner tensor is None, you should not return an attribute for it).

```
import torch

class OptionalScaledTensor(torch.Tensor):
    def __new__(
        cls,
        data,
        scale,
        *,
        constant: int = 0,
    ):
        return torch.Tensor._make_wrapper_subclass(
            cls,
            data.size(),
            strides=data.stride(),
            storage_offset=data.storage_offset(),
            dtype=data.dtype,
            layout=data.layout,
            requires_grad=data.requires_grad,
            device=data.device,
        )

    def __init__(self, data: torch.Tensor, scale, constant: int = 0):
        self._data = data
        self._scale = scale
        self._constant = constant

    def __tensor_flatten__(self):
        ctx = {"_constant": self._constant}
        if self._scale is not None:
            return ["_data", "_scale"], ctx
        else:
            return ["_data"], ctx

    @staticmethod
    def __tensor_unflatten__(inner_tensors, metadata, outer_size, outer_stride):
        return OptionalScaledTensor(
            inner_tensors["_data"],
            inner_tensors["_scale"] if '_scale' in inner_tensors else None,
            constant=metadata["_constant"],
        )

    @classmethod
    def __torch_dispatch__(cls, func, types, args, kwargs=None):
        scaled_tensor = args[0]
        out = func(scaled_tensor._data, *args[1:], **kwargs)
        if scaled_tensor._scale is not None:
            out = out * scaled_tensor._scale
        return OptionalScaledTensor(out, scaled_tensor._scale, constant=scaled_tensor._constant)

    def __repr__(self):
        return f"OptionalScaledTensor({self._data.__repr__()}\n{self._scale.__repr__()})"


def f(x):
    return x + 1

sub1 = OptionalScaledTensor(torch.randn(2, 4), None)
sub2 = OptionalScaledTensor(torch.randn(2, 4), torch.randn(2, 4))

f_compiled = torch.compile(f, backend='aot_eager')

# compile with a none scale
f_compiled(sub1)
# this should recompile, scale is not None
out_test = f_compiled(sub2)
out_ref = f(sub2)
print(torch.allclose(out_test._data, out_ref._data))
```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @albanD @anijain2305 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129618

Fixes #129601

Background: it's possible that a traceable wrapper subclass will have an optional inner tensor constituent (e.g. NJT's cached min / max sequence lengths). To specify this, the subclass's `__tensor_flatten__()` impl should leave out any unspecified optional inner tensors in the returned list of `attrs`.

This PR guards on the list of inner tensor `attrs` returned in `subclass.__tensor_flatten__()[0]`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

**Review by anijain2305 — APPROVED** [review]

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -1396,6 +1397,10 @@ def wrap_tensor(self, value: torch.Tensor):
         # and recursively install corresponding guard for each inner attribute.
         if is_traceable_wrapper_subclass(value):
             self.install_guards(GuardBuilder.TYPE_MATCH)
+            install_guard(
+                SubclassAttrListSource(source).make_guard(GuardBuilder.EQUALS_MATCH)
```
Comment (bdhirsh) [inline torch/_dynamo/variables/builder.py:1401]: cc @tugsbayasgalan - we'll need to fix this for nested subclasses too. Dynamo generate a type match (and also am match on tensor_flatten attributes), and we should generate these guards recursively for inner tensors. (We can fix this independently of the metadata guard [issue](https://github.com/pytorch/pytorch/issues/114405))

**Pull request closed without merging**